app/src/main/java/app/revanced/manager/ui/screen/PatcherScreen.kt

```diff
@@ -0,0 +1,433 @@
+                            steps.all { it.state == State.WAITING } -> context.getString(R.string.apply_patches)
+                            else -> context.resources.getQuantityString(R.plurals.applied_patches, steps.size, steps.size)
+                        }
+                    }
+
+                    val state = remember(steps) {
+                        when {
+                            steps.all { it.state == State.COMPLETED } -> State.COMPLETED
+                            steps.any { it.state == State.FAILED } -> State.FAILED
+                            steps.any { it.state == State.RUNNING } -> State.RUNNING
+                            else -> State.WAITING
+                        }
+                    }
+
+                    SubStep(
+                        name = name,
+                        state = state,
+                        message = steps.find { it.state == State.FAILED }?.message
+                    )
+                }
+            }
+        }
+    }
+}
+
+@Composable
+fun SubStep(
+    name: String,
+    state: State,
+    message: String? = null,
+    downloadProgress: Pair<Float, Float>? = null
+) {
+    var messageExpanded by rememberSaveable { mutableStateOf(true) }
+
+    Column(
+        modifier = Modifier
+            .run {
+                if (message != null)
+                    clickable { messageExpanded = !messageExpanded }
+                else this
+            }
+    ) {
+        Row(
+            verticalAlignment = Alignment.CenterVertically,
+            horizontalArrangement = Arrangement.spacedBy(12.dp),
+            modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp)
+        ) {
+            Box(
+                modifier = Modifier.size(24.dp),
+                contentAlignment = Alignment.Center
+            ) {
+                StepIcon(state, downloadProgress, size = 20.dp)
+            }
+
+            Text(
+                text = name,
+                style = MaterialTheme.typography.titleSmall,
+                maxLines = 1,
+                overflow = TextOverflow.Ellipsis,
+                modifier = Modifier.weight(1f, true),
+            )
+
+            if (message != null) {
+                Box(
+                    modifier = Modifier.size(24.dp),
+                    contentAlignment = Alignment.Center
+                ) {
+                    ArrowButton(
+                        modifier = Modifier.size(20.dp),
+                        expanded = messageExpanded,
+                        onClick = null
+                    )
+                }
+            } else {
+                downloadProgress?.let { (current, total) ->
+                    Text(
+                        "$current/$total MB",
+                        style = MaterialTheme.typography.labelSmall
+                    )
+                }
+            }
+        }
```

Comment:
Why is download progress handled in the SubStep component? if necessary, a parameter for extra information can be exposed

Reply:
Because we would have to add another StateFlow to the substep while we only use it for download progress right now.